File: author_backend_nodejs/src/api/exercises.js
'use strict';
var fs = require('mz/fs');
var xml2js = require('xml2js');
var xmlparser = new xml2js.Parser({trim:true});
var _ = require('lodash');
import { compareExpressions, toLatex } from './symbolic.js'; 


function consoleJSON(obj) {
  console.log(JSON.stringify(obj, null, 4));
}

function parseXML(xml) {
  return new Promise( (resolve, reject) => {
    xmlparser.parseString(xml, (err, result) => {
      if(err) reject(err);
      resolve(result);
    });
  });
}
// () => List
// List available exercises
export function listExercises() {
  return fs.readdir("./exercises/").catch( err => console.log(err) );
}

export function getExerciseXMLasJSON(exercise) {
  var parser = new xml2js.Parser();
  var problemdata = fs.readFile(exercise, 'utf8')
    //.then( file => { console.log(file); return file;} )
    .then( parseXML )
    //.then( data => { consoleJSON(data); return data; } )
    .catch( err => console.log(exercise + ': ' + err) );
  return problemdata;
}

function parseIngress(ingress) {
  try{ 
    var variables = ingress.split(';')
  .filter( x => x !== "" )
  .map( x => x.split('=') )
  .map( x => ({ name: x[0].trim(), value: x[1].trim() }) );
  return variables;
  } catch(err) {
    console.dir(err);
    return {}
  }
}

export function checkQuestion(exercise, question, expression) {
  return getExerciseXMLasJSON('./exercises/' + exercise + '/problem.xml')
  .then(json => {
    if(_.has(json,'problem.thecorrectanswer[' + question + ']')) {
      var correct = json.problem.thecorrectanswer[question]._
      .replace(';', '');
      //Parse ingress
      try {
        var variables = parseIngress(json.problem.thecorrectanswer[0]._);
      console.log(expression);
        return Promise.all(
          [
            compareExpressions(JSON.stringify(variables), correct, expression),
            toLatex(expression)
          ])
          .then( res => ({ equal: res[0], latex: res[1] }) );
      } catch(err) {
        console.dir(err)
        return {
          error: err
        }
      }
    }
    else {
      return {error: 'Cannot extract question ' + question + ' from exercise ' + exercise};
    }
  });
}

File: pyramid/src/exercises.py
from pyramid.response import Response, FileResponse
from xmljson import badgerfish as bf
import traceback
from xml.etree.ElementTree import fromstring
import os
import sys
import functools
import operator
import json as JSON
from symbolic_server import Symbolic
from time import sleep
from random import random

symbolic = Symbolic()


def deep_get(dictionary, *keys):
    return functools.reduce(lambda d, key: d.get(key) if d else None, keys, dictionary)


def nested_print(d):
    print(JSON.dumps(d, indent=4))
    return


def exercises():  # {{{
    exerciselist = []
    try:
        exerciselist = [
            name
            for name in os.listdir('./exercises/')
            if os.path.isdir(os.path.join('./exercises', name))
        ]
    except Exception:
        print(traceback.format_exc())
        pass
    print(exerciselist)
    return exerciselist  # }}}


# Need to split into JSON reading part and request handling
def exerciseJSON(path):  # {{{
    obj = {}
    try:
        print('./exercises/{path}/problem.xml'.format(path=path))
        xmlfile = open('./exercises/{path}/problem.xml'.format(path=path))
        xml = xmlfile.read()
        obj = bf.data(fromstring(xml))
    except Exception:
        print("exerciseJSON: Error reading or parsing JSON for {path}".format(path=path))
        print(traceback.format_exc())
        pass
    return obj  # }}}


def exerciseXML(path):  # {{{
    xml = ''
    try:
        xmlfile = open('./exercises/{path}/problem.xml'.format(path=path))
        xml = xmlfile.read()
    except Exception:
        print("exerciseXML: Error reading XML for {path}".format(path=path))
        print(traceback.format_exc())
        pass
    return xml  # }}}


def compose(*funcs):
    return lambda x: functools.reduce(lambda v, f: f(v), funcs, x)


def parseIngress(ingress):
    rawvars = ingress.split(';')
    pipeline = compose(
        functools.partial(filter, operator.truth),
        functools.partial(map, lambda x: x.split('=')),
        functools.partial(map, lambda x: {'name': x[0].strip(), 'value': x[1].strip()}),
    )
    variables = list(pipeline(rawvars))
    return variables


def exerciseCheck(exercise, question, expression):  # {{{
    json = exerciseJSON(exercise)
    # print(JSON.dumps(deep_get(json,'problem','thecorrectanswer'), indent=4))
    questions = deep_get(json, 'problem', 'thecorrectanswer')
    if question < len(questions):
        variables = parseIngress(questions[0].get('$'))
        correct = questions[question].get('$').replace(';', '')
        result = symbolic.compareNumeric(JSON.dumps(variables), expression, correct)
        latex = {'latex': symbolic.toLatex(expression)}
        result.update(latex)
        # Need to merge with result dictionary...
        return result

        # nested_print(questions[question].get('$'))
    # nested_print(questions[1].get('$'))

    return True  # }}}


def exerciseSave(exercise, xml):
    print('Saving ' + exercise)
    with open('./exercises/{path}/problem.xml'.format(path=exercise), 'w') as file:
        file.write(xml)
    sleep(0.5)
    if random() > 0.5:
        raise IOError('Simulated IOError')
    return {'success': True}


# This is waiting for rewrite of exerciseJSON
# def checkExercise(exercise, question, expression):
#    json = exer
